File: java/partitionlist.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode partition(ListNode head, int x) {
        // Start typing your Java solution below
        // DO NOT write main() function
        
        ListNode head1 = new ListNode(0);
        ListNode head2 = new ListNode(0);
        ListNode tail1 = head1;
        ListNode tail2 = head2;
        
        ListNode p = head;
        while(p!=null){
            if(p.val<x){
                tail1.next = p;
                tail1 = tail1.next;
            }
            else{
                tail2.next = p;
                tail2 =tail2.next;
            }
            p = p.next;
        }
        
        tail1.next = head2.next;
        tail2.next = null;
        return head1.next;
        
    }
}

File: cpp/PartitionList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *partition(ListNode *head, int x) {
        ListNode* smaller = new ListNode(0);
        ListNode* pSmaller = smaller;
        ListNode* bigger = new ListNode(0);
        ListNode* pBigger = bigger;
        ListNode* p = NULL;
        while(head!=NULL) {
            p = head;
            head = head->next;
            p->next = NULL;
            
            if(p->val < x) {
                pSmaller->next = p;
                pSmaller = p;
            } else {
                pBigger->next = p;
                pBigger = p;
            }
        }
        pSmaller->next = bigger->next;
        
        ListNode* res = smaller->next;
        if(res == NULL) {
            res = bigger->next;
        }
        return res;
    }
};
